File: backend/services/lcl_fees.py
"""
LCL Bourse fee simulation — tarifs publics LCL 2024.
Sources: grille tarifaire LCL Bourse (ordres internet, compte-titres ordinaire).
"""

from dataclasses import dataclass

# Seuils courtage internet LCL Bourse (ordres sur marchés règlementés)
COURTAGE_TRANCHES = [
    (800,   8.50,  None),   # ≤ 800€ → forfait 8.50€
    (3000,  None,  0.0070), # 800–3000€ → 0.70 %
    (float("inf"), 15.00, 0.0050),  # > 3000€ → 0.50 % min 15€
]

# Taxe sur les Transactions Financières (PTF/TTF) — art. 235 ter ZD CGI
# 0.3 % sur achats d'actions françaises cotées si cap. bours. > 1 Mrd€
TTF_RATE = 0.003

# SRD — report en date de liquidation
SRD_RATE = 0.00032  # 0.032 % du montant nominal par report

# Droits de garde annuels (simulés par ordre, base annuelle)
DROITS_GARDE_RATE = 0.0012  # 0.12 % / an, min 3 € par ligne
DROITS_GARDE_MIN  = 3.00

# Droits de timbre UK (stamp duty) pour London Stock Exchange
STAMP_DUTY_UK_RATE = 0.005  # 0.5 %


@dataclass
class FeeBreakdown:
    montant_brut: float
    courtage: float
    ttf: float
    srd: float
    droits_garde_annuels: float
    total_frais: float
    montant_net_achat: float
    montant_net_vente: float
    taux_effectif_pct: float


def _courtage(montant: float) -> float:
    for seuil, forfait, taux in COURTAGE_TRANCHES:
        if montant <= seuil:
            if forfait is not None:
                return forfait
            fee = montant * taux
            return max(fee, COURTAGE_TRANCHES[2][1])  # min 15€ pour dernière tranche
    return montant * COURTAGE_TRANCHES[-1][2]


def simuler_ordre(
    montant: float,
    marche: str = "Euronext Paris",
    action_francaise: bool = True,
    eligible_ttf: bool = True,
    srd: bool = False,
) -> FeeBreakdown:
    """
    Calcule le détail des frais LCL Bourse pour un ordre.

    Args:
        montant: montant de l'ordre en euros (quantité × cours)
        marche: place de cotation
        action_francaise: True si l'action est cotée en France
        eligible_ttf: True si cap. bours. > 1 Mrd€ (TTF applicable)
        srd: True si l'ordre est passé avec SRD (report)
    """
    courtage = _courtage(montant)

    ttf = 0.0
    if action_francaise and eligible_ttf:
        ttf = montant * TTF_RATE

    srd_frais = montant * SRD_RATE if srd else 0.0

    garde = max(montant * DROITS_GARDE_RATE / 12, DROITS_GARDE_MIN / 12)  # mensuel

    total = courtage + ttf + srd_frais
    net_achat = montant + total
    net_vente = montant - total
    taux_eff = (total / montant) * 100 if montant > 0 else 0

    return FeeBreakdown(
        montant_brut=round(montant, 2),
        courtage=round(courtage, 2),
        ttf=round(ttf, 2),
        srd=round(srd_frais, 2),
        droits_garde_annuels=round(montant * DROITS_GARDE_RATE, 2),
        total_frais=round(total, 2),
        montant_net_achat=round(net_achat, 2),
        montant_net_vente=round(net_vente, 2),
        taux_effectif_pct=round(taux_eff, 4),
    )


def seuil_rentabilite(prix_achat: float, frais_achat: float, frais_vente_rate: float = 0.007) -> float:
    """Prix minimum de revente pour couvrir les frais aller-retour."""
    return prix_achat + frais_achat / 1 + (prix_achat * frais_vente_rate)

File: frontend/src/utils/lcl-fees.ts
/**
 * Calcul des frais LCL Bourse — grille tarifaire 2024
 * Ordres internet, compte-titres ordinaire (pas PEA, pas SRD)
 */

/**
 * Détermine si une valeur est éligible au PEA français.
 * Règle simplifiée (applicable pour LCL PEA) :
 *   - Actions de sociétés domiciliées dans l'UE / EEE sont éligibles
 *   - UK (.L) exclue depuis le Brexit
 *   - USA, Japon, etc. exclus
 *   - ETF : seuls les ETF UCITS domiciliés dans l'UE sont éligibles
 *     (on ne peut pas le savoir depuis le symbole seul → false par défaut)
 */
export function isPeaEligible(symbol: string): boolean {
  const s = symbol.toUpperCase()
  // Suffixes Euronext / marchés UE+EEE
  const EU_SUFFIXES = [
    '.PA',  // Euronext Paris (France)
    '.DE',  // XETRA (Allemagne)
    '.AS',  // Euronext Amsterdam (Pays-Bas)
    '.BE',  // Euronext Bruxelles (Belgique)
    '.MI',  // Borsa Italiana (Italie)
    '.MC',  // Bolsa de Madrid (Espagne)
    '.LS',  // Euronext Lisbonne (Portugal)
    '.HE',  // Helsinki (Finlande)
    '.ST',  // Stockholm (Suède)
    '.CO',  // Copenhague (Danemark)
    '.OL',  // Oslo (Norvège)
    '.VI',  // Vienne (Autriche)
    '.WA',  // Varsovie (Pologne)
    '.PR',  // Prague (Tchéquie)
    '.BD',  // Budapest (Hongrie)
  ]
  return EU_SUFFIXES.some(sfx => s.endsWith(sfx))
}

export interface LclFeeBreakdown {
  courtage:     number   // frais de courtage
  ttf:          number   // taxe sur transactions financières (0.3% actions FR)
  stampDuty:    number   // stamp duty UK (0.5% actions .L)
  total:        number   // total frais pour cet ordre (achat)
  totalAller:   number   // frais achat (= total)
  tauxEffectif: number   // % du montant brut
  label:        string   // description lisible
}

/** Courtage internet LCL sur le montant de l'ordre */
function calcCourtage(montant: number): number {
  if (montant <= 800)  return 8.50
  if (montant <= 3000) return montant * 0.007
  return Math.max(montant * 0.005, 15)
}

/**
 * Détermine si la TTF (0.30%) s'applique.
 * S'applique aux achats d'actions françaises avec cap. boursière > 1 Mrd€.
 * On considère tous les titres Euronext Paris (.PA) comme éligibles par défaut
 * (hors micro-caps — mais conservatif = on l'applique toujours pour .PA).
 */
function isTtfEligible(symbol: string): boolean {
  return symbol.toUpperCase().endsWith('.PA')
}

/** Stamp duty UK (0.5% à l'achat uniquement) */
function isUkStamp(symbol: string): boolean {
  return symbol.toUpperCase().endsWith('.L')
}

/**
 * Calcule les frais LCL pour un ordre d'achat.
 * @param symbol   Symbole Yahoo Finance (ex: MC.PA, AAPL, SHEL.L)
 * @param qty      Quantité d'actions
 * @param price    Prix unitaire en €
 */
export function calcLclFees(symbol: string, qty: number, price: number): LclFeeBreakdown {
  const montant  = qty * price
  if (!montant || montant <= 0) {
    return { courtage: 0, ttf: 0, stampDuty: 0, total: 0, totalAller: 0, tauxEffectif: 0, label: '' }
  }

  const courtage  = round2(calcCourtage(montant))
  const ttf       = isTtfEligible(symbol)  ? round2(montant * 0.003)  : 0
  const stampDuty = isUkStamp(symbol)      ? round2(montant * 0.005)  : 0
  const total     = round2(courtage + ttf + stampDuty)
  const tauxEffectif = montant > 0 ? round4(total / montant * 100) : 0

  const parts: string[] = [`courtage ${courtage.toFixed(2)} €`]
  if (ttf > 0)       parts.push(`TTF ${ttf.toFixed(2)} €`)
  if (stampDuty > 0) parts.push(`stamp duty ${stampDuty.toFixed(2)} €`)

  return {
    courtage,
    ttf,
    stampDuty,
    total,
    totalAller:   total,
    tauxEffectif,
    label: parts.join(' + '),
  }
}

/** Seuil de rentabilité : prix de revente minimum pour couvrir les frais A/R */
export function calcBreakeven(buyPrice: number, qty: number, fees: number): number {
  if (!qty || !buyPrice) return 0
  // On estime les frais de vente = même courtage sur même montant
  const montantAchat = buyPrice * qty
  const fraisVente   = calcCourtage(montantAchat)
  return round2(buyPrice + (fees + fraisVente) / qty)
}

function round2(v: number) { return Math.round(v * 100) / 100 }
function round4(v: number) { return Math.round(v * 10000) / 10000 }
